contracts/actions/mcd/McdRepayComposite.sol

```diff
@@ -0,0 +1,200 @@
+    ) public payable virtual override(ActionBase, DFSSell, GasFeeTaker) returns (bytes32) {
+        RepayParams memory repayParams = _parseCompositeParams(_callData);
+
+        repayParams.vaultId = _parseParamUint(
+            repayParams.vaultId,
+            _paramMapping[0],
+            _subData,
+            _returnValues
+        );
+
+        repayParams.joinAddr = _parseParamAddr(
+            repayParams.joinAddr,
+            _paramMapping[1],
+            _subData,
+            _returnValues
+        );
+
+        repayParams.exchangeData.srcAddr = _parseParamAddr(
+            repayParams.exchangeData.srcAddr,
+            _paramMapping[2],
+            _subData,
+            _returnValues
+        );
+        repayParams.exchangeData.destAddr = _parseParamAddr(
+            repayParams.exchangeData.destAddr,
+            _paramMapping[3],
+            _subData,
+            _returnValues
+        );
+
+        repayParams.exchangeData.srcAmount = _parseParamUint(
+            repayParams.exchangeData.srcAmount,
+            _paramMapping[4],
+            _subData,
+            _returnValues
+        );
+
+        (bytes memory logData, uint256 paybackAmount) = _repay(repayParams);
+        emit ActionEvent("McdRepayComposite", logData);
+
+        return bytes32(paybackAmount);
+    }
+
+    /// @inheritdoc ActionBase
+    function executeActionDirect(bytes memory _callData)
+        public
+        payable
+        virtual
+        override(ActionBase, DFSSell, GasFeeTaker)
+    {
+        RepayParams memory repayParams = _parseCompositeParams(_callData);
+        (bytes memory logData, ) = _repay(repayParams);
+        logger.logActionDirectEvent("McdRepayComposite", logData);
+    }
+
+    /// @notice Executes repay logic
+    function _repay(RepayParams memory _repayParams) internal returns (bytes memory logData, uint256 paybackAmount) {
+        assert(_repayParams.exchangeData.destAddr == DAI_ADDR);
+        
+        uint256 ratioBefore;
+        // is part of strategy so check before ratio
+        if (_repayParams.gasUsed != 0) {
+            ratioBefore = getRatio(_repayParams.vaultId, 0);
+        }
+
+        uint256 repayAmount = _repayParams.exchangeData.srcAmount;
+
+        // Withdraw collateral if not using flashloan
+        if (_repayParams.flAddress == address(0)) {
+            _withdrawColl(_repayParams, repayAmount);
+        }
+
+        // Sell collateral asset for debt asset
+        (uint256 exchangedAmount, ) = _dfsSell(
+            _repayParams.exchangeData,
+            address(this),
+            address(this),
+            false
+        );
+
+        // Take gas fee if part of strategy
+        if (_repayParams.gasUsed != 0) {
+            paybackAmount = _takeFee(
+                GasFeeTakerParams(_repayParams.gasUsed, DAI_ADDR, exchangedAmount, MAX_DFS_FEE)
+            );
+        } else {
+            paybackAmount = exchangedAmount;
+        }
+
+        // Payback debt
+        {
```

Comment:
maybe put this in helper function

Reply:
less gas efficient, that's why it's all block scoped